Source/Views/TableViews/SPCopyTable.m

```diff
@@ -448,49 +448,58 @@ - (NSString *)rowsAsSqlInsertsOnlySelectedRows:(BOOL)onlySelected skipAutoIncrem
 		columnMappings[c] = (NSUInteger)[[[columns safeObjectAtIndex:c] identifier] integerValue];
 		
 		NSString *t = [[columnDefinitions safeObjectAtIndex:columnMappings[c]] objectForKey:@"typegrouping"];
-
-		if(foundAutoIncColumn == NO && skipAutoIncrementColumn == YES){
-            
-            id obj = [columnDefinitions safeObjectAtIndex:columnMappings[c]];
-                        
-            if ([obj respondsToSelector:@selector(boolForKey:)]) {
-                autoIncrement = [obj boolForKey:@"autoincrement"];
-                // the columnDefinitions array contains dictionaries with different keys when copying from the table view (autoincrement)
-                // or the query view (AUTO_INCREMENT_FLAG)
-                // so we need this extra check
-                if(autoIncrement == NO){
-                    autoIncrement = [obj boolForKey:@"AUTO_INCREMENT_FLAG"];
+        NSString *g = [[columnDefinitions safeObjectAtIndex:columnMappings[c]] objectForKey:@"generatedalways"];
+
+        // Generated column
+        if (skipGeneratedColumn == YES && g) {
+            [tbHeader removeObject:[[columnDefinitions safeObjectAtIndex:columnMappings[c]] objectForKey:@"name"]];
+            columnTypes[c] = 5;
+        } else {
+        // AutoInc column
```

Comment:
I think removing the column from the tbHeader here can introduce errors below with the tbHeader:safeObjectAtIndex call for later columns.
Why do you need to call removeObject here at all? I believe setting the column type alone should be sufficient to create the desired behavior?

Reply:
Removing the column from the tbHeader is required here (Line 541-543): 

```
    // Begin the SQL string
    [result appendFormat:@"INSERT INTO %@ (%@)\nVALUES\n",
        [(selectedTable == nil) ? @"<table>" : selectedTable backtickQuotedString], [tbHeader componentsJoinedAndBacktickQuoted]];
```

It should not generate an error, since we are using a pointer, not an index to reference the object in the array. More over the safeObjectAtIndex method is safer than ObjectAtIndex.

Reply:
When you call removeObject on an NSMutableArray (tbHeader), that should cause indexes after the generated column to jump up by 1, correct? So, if the generated column(s) aren't at the end of the table, that later call (`[tbHeader safeObjectAtIndex:columnMappings[c]]` around 576) will result in the wrong table header being returned for columns in the table that come after the generated columns, correct?
In other words, the safe object for index thing isn't the issue, the issue is the shift in array indexes.
If you can test and verify that the correct insert statement is actually generated for tables that have non-generated columns after generated columns, then I'll feel better about this PR but I don't see how it would work correctly in that case

Reply:
We could convert tbHeader into a dictionary perhaps and then not have issues with re-keying?
Perhaps I'm not reading the code correctly also, but it really seems like it would break down in the case described above.

Overall, the whole function is sorta hacky. Perhaps it could be refactored further to be simpler down the road.

Reply:
You're right, `[tbHeader safeObjectAtIndex:columnMappings[c]]` around 576 could be an issue. It occurs also when auto increment is removed (`[tbHeader removeObject:autoIncrementColumnName];`around 524.
But the error only gets thrown when we try to reload data and in my testing that part of code was never used. But, historically, I think if it's here it must be useful ?!

I'm agree the whole function is not very clean. I'll try to refactored it !

Reply:
I don't have a general idea of how to make it better, except maybe it's worth using dictionaries/objects instead of having multiple arrays and using indexes to keep them all in sync :)

Reply:
@gduh Any progress on this fix? I'd love to try to get it out in a release soon!

Reply:
> @gduh Any progress on this fix? I'd love to try to get it out in a release soon!

@Jason-Morcos  I'm trying to push this fix before the end of this week, ok for you ?

Reply:
Works for me! Hoping to push a release with a bunch of fixes very beginning of 2022. Thank you for your work!